File: Hashmap/Ransomenote.py
'''
Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.

Each letter in magazine can only be used once in ransomNote.
'''
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        # Count frequency of each character in magazine
        magazine_count = {}
        for char in magazine:
            magazine_count[char] = magazine_count.get(char, 0) + 1
        
        # Check if we can construct ransomNote
        for char in ransomNote:
            if magazine_count.get(char, 0) == 0:
                return False
            magazine_count[char] -= 1
        
        return True

File: Hashmap/Ransomenote.rs
use std::collections::HashMap;

/*
Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.

Each letter in magazine can only be used once in ransomNote.
*/
impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        let mut magazine_count = HashMap::new();
        for ch in magazine.chars() {
            *magazine_count.entry(ch).or_insert(0) += 1;
        }
        
        for ch in ransom_note.chars() {
            let count = magazine_count.entry(ch).or_insert(0);
            if *count == 0 {
                return false;
            }
            *count -= 1;
        }
        
        true
    }
}